ASoC: Intel: sof_sdw: run rtd_init for each dai in a link

bardliao opened the pull request:
The existing code assumes that all codecs in a dai link are the same model. However, people may use different amp module in a dai link. This series adds a rtd_init callback to sof_sdw_dai_info{} and call the rtd_init callback in a common dai_link .init to initialize each codec in the dai link.

bardliao committed:
ASoC: Intel: sof_sdw_rt_sdca_jack_common: use helper to get codec dai by
name

Use helper to get codec dai by name.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: add get_codec_dai_by_name helper function

Currently, we assume the codecs in a dai link are all the same.
So that we get codec dai with snd_soc_rtd_to_codec(rtd, 0) in dai_links
->init callback. However, a link can include different codecs.
For example, a 4 speakers link can consist of rt712 and rt1316.
Therefore, we need to select the codec dai by name in the dai link.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw_rt711: use helper to get codec dai by name

Use helper to get codec dai by name.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw_rt712_sdca: use helper to get codec dai by name

Use helper to get codec dai by name.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw_rt700: use helper to get codec dai by name

Use helper to get codec dai by name.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw_cs42l42: use helper to get codec dai by name

Use helper to get codec dai by name.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw_rt5682: use helper to get codec dai by name

Use helper to get codec dai by name.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw: use single rtd_init for rt_amps

2 amps can be in the same or different dai links. To handle this, the
existing code implements different spk_init functions to add dapm routes
for different amps. However, sof_sdw.c doesn't support non-aggregated
amp any more since it used pre-defined BE id.
With that assumption, combine the spk_init functions together.
This is a preparation of putting different types amps in a single dai
link.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw: add common sdw dai link init

Currently, we set sdw dai link .init callback in the codec_info_list's
dais.init function. This works fine if all codecs in the dai link are
the same. However, we need to do all the .init stuff for all different
codecs in the dai link if not all codecs in the dai link are the same.
Use a common dai link .init callback to call the new rtd_init callback
in sof_sdw_dai_info{} to do rtd_init for each dai.
Some codec init callback will become empty after this change. They will
be removed in the follow up patch.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
A remaining open is that
`SOC_DAPM_PIN_SWITCH("Speaker"),`
`SND_SOC_DAPM_SPK("Speaker", NULL),`
will be added twice if rt712 and rt1316 are in the same dai link.
[sof_sdw_rt_amp.c](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw_rt_amp.c#L203)
[sof_sdw_rt712_sdca.c](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw_rt712_sdca.c#L48)

plbossart commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -865,7 +872,7 @@ static struct sof_sdw_codec_info codec_info_list[] = {
 				.dai_type = SOF_SDW_DAI_TYPE_MIC,
 				.dailink = {SDW_UNUSED_DAI_ID, SDW_DMIC_DAI_ID},
-				.init = sof_sdw_rt715_init,
+				.rtd_init = rt715_rtd_init,
```
I couldn't follow this patch. In some cases, .init is removed and seems to be replaced by .rtd_init, but...

bardliao replied:
[codec_info_list[codec_index].dais[dai_index].init](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c#L1318) is called when the machine driver is creating dai links. And the `.rtd_init` will be called in the dai_link .init callback. And some codec_info_list `.init` functions only does one thing which is assign the `dai_links->init` function.
Like
```
int sof_sdw_rt715_init(struct snd_soc_card *card,
		       const struct snd_soc_acpi_link_adr *link,
		       struct snd_soc_dai_link *dai_links,
		       struct sof_sdw_codec_info *info,
		       bool playback)
{
	dai_links->init = rt715_rtd_init;

	return 0;
}
```
And the rtd_init is already called by `sof_sdw_rtd_init`. So the .init function is not needed any more.
Let me separate this to two patches. It may be easier to review.

plbossart commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -892,6 +900,7 @@ static struct sof_sdw_codec_info codec_info_list[] = {
 				.dai_type = SOF_SDW_DAI_TYPE_AMP,
 				.dailink = {SDW_AMP_OUT_DAI_ID, SDW_UNUSED_DAI_ID},
 				.init = sof_sdw_maxim_init,
+				.rtd_init = maxim_spk_rtd_init,
```
... here we have both callbacks. I really can't follow which callback does what.

plbossart reviewed (commented):
very hard to follow @bardliao

bardliao committed:
ASoC: Intel: sof_sdw: remove .init callbacks

Some codec .init callbacks are empty after removing dai_links->init =
xxx_rtd_init;. Remove those callbacks.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -682,6 +683,7 @@ static struct sof_sdw_codec_info codec_info_list[] = {
 				.dailink = {SDW_JACK_OUT_DAI_ID, SDW_JACK_IN_DAI_ID},
 				.init = sof_sdw_rt_sdca_jack_init,
 				.exit = sof_sdw_rt_sdca_jack_exit,
+				.rtd_init = rt_sdca_jack_rtd_init,
```
sorry, I am not following what the change is.

we had a perfectly symmetric solution with init/exit, but now init is replaced by rtd_init, should we have an rtd_exit?

This looks rather odd for now, I don't know what to make of this.

bardliao replied:
> sorry, I am not following what the change is.
> 
> we had a perfectly symmetric solution with init/exit, but now init is replaced by rtd_init, should we have an rtd_exit?


The init/exit was not perfectly symmetric for example, there is a [.init = sof_sdw_rt700_init,](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c#L669), and no sof_sdw_rt700_exit. And the only thing [sof_sdw_rt700_init](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw_rt700.c#L113) does is to assign `dai_links->init = rt700_rtd_init;`.
The PR removes `dai_links->init = xxx_rtd_init;` from the .init function, and run `xxx_rtd_init` in `sof_sdw_rtd_init`. 
We don't have dai_link.exit callback for now. But, yes, we probably, need a `sof_sdw_rtd_exit` to call rtd_exit in case some codecs need it someday.

> 
> This looks rather odd for now, I don't know what to make of this.

plbossart commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -1371,6 +1393,55 @@ static void set_dailink_map(struct snd_soc_dai_link_codec_ch_map *sdw_codec_ch_m
+		codec_info = &codec_info_list[codec_index];
+		/*
+		 * A codec dai can be connected to different dai links for capture and playback,
+		 * but we only need to call the rtd_init function once.
```
this is super confusing. The rtd_init function uses the 'rtd' parameter, so could it be that the rtd_init() behaves differently depending on the order in which dais are declared.

This goes back to the other comment on the 'initialized' field, we should try to make this a bit cleaner IMHO.

bardliao replied:
The rtd_init will be called based on the order of codec dais. However, the order does not really matter. Typically, we append card->components, add card controls, dapm controls, dapm routes in rtd_init functions.  It doesn't matter if we add A first or B first. But, I do have an open that two rtd_init may add controls with the same name. For example, rt712 and rt1316 both add `SOC_DAPM_PIN_SWITCH("Speaker")`. 
Regarding the 'initialized' field, maybe rename to `rtd_init_done`, and add a comment to describe it?

plbossart replied:
oh, ok. that would work if the codec dais are indeed on the same dailink. Otherwise I don't see how this would work since the same dai would be used for different purposes, and then the rtd_init_done would make no sense.

Do I get this right now? If yes, it's worth adding a big fat comment on what this rtd_init_done covers and when it would not work.

bardliao replied:
Sorry, @plbossart I didn't get your point of "if the codec dais are indeed on the same dailink". I use `snd_soc_rtd_to_codec` here to go through each codec dai in a dailink. Suppose all the codec dais are in the same dailink, right?
Or do you mean for example, `rt712-sdca-aif1` dai will be in both `SDW0-Playback` and `SDW0-Capture` dailinks?
The purpose of `rtd_init_done` is to make sure a dai's rtd_init callback will be called once. For example, `rt712-sdca-aif1`'s rtd_init will be called be the SDW0-Playback` dailink. And when `SDW0-Capture` dailink call `rt712-sdca-aif1`'s rtd_init again, it will `continue`.
It will also work when `rt1316-aif` dais in the same dailink. The second `rt1316-aif` dai's rtd_init will not be called as `rt1316-aif`'s rtd_init_done will be true at that moment.
The behavior matches the existing behavior. The existing dai.init function will set `dai_links->init = xxx_rtd_init` in just one dailink. For example,  `sof_sdw_rt_sdca_jack_init` will return 0 and do nothing `if (!playback)`. And so do other dai.init functions.

plbossart replied:
it goes back to the change we made: all instances of a dai need to be in the same rtd/dailink.

We don't support e.g. two headset codecs used independently, or two identical amplifiers handling different content on separate links.

bardliao replied:
> it goes back to the change we made: all instances of a dai need to be in the same rtd/dailink.
> 
> We don't support e.g. two headset codecs used independently, or two identical amplifiers handling different content on separate links.

No, we don't. The only case a dai can be in different dailinks is that it is in a capture dailink and a playback dailink.
For example, `SDW0-Playback` and `SDW0-Capture` dailinks both uses `rt711-sdca-aif1`.

Let me explain the PR with some examples.

The existing code assumes all codecs in a dailink are the same model.
So, we set dai_link.init = xxx_rtd_init in the codec dai.init function. For example, 
Say we have rt711-sdca jack and rt1316 amp.
`SDW0-Playback` dailink .init = rt_sdca_jack_rtd_init
`SDW0-Capture` dailink .init = NULL    #because `sof_sdw_rt_sdca_jack_init` will only set dai_link.init = rt_sdca_jack_rtd_init when playback is true
`SDW1-Playback` dailink .init = all_spk_init
`SDW1-Capture` dailink .init = NULL #because `sof_sdw_rt_amp_init` will only set dai_link.init = sof_sdw_rt_amp_init when playback is true

The PR is trying to remove the restriction of using the same codec model in a dailink, so, we need to call the xxx_rtd_init functions for different codec models in a dailink.
With the change, the dailink.init are all `sof_sdw_rtd_init`, and `sof_sdw_rtd_init` will call xxx_rtd_init for each codec model in the dailink. Use the same example, we will have
`SDW0-Playback` dailink .init = sof_sdw_rtd_init which will call rt_sdca_jack_rtd_init and set `rt711-sdca-aif1` dai.rtd_init_done = true
`SDW0-Capture` dailink .init = sof_sdw_rtd_init which will call nothing since `rt711-sdca-aif1` dai.rtd_init_done is true.
`SDW1-Playback` dailink .init = sof_sdw_rtd_init which will call `rt_amp_spk_rtd_init` once since `rt1316-aif` dai.rtd_init_done will be true after the first `rt_amp_spk_rtd_init` is called.
`SDW1-Capture` dailink .init = sof_sdw_rtd_init which will call nothing since `rt1316-aif` dai.rtd_init_done is true.

If we have rt712-sdca jack on `SDW0-Playback` and `SDW0-Capture` dailinks and rt712-sdca amp + rt1316 amp on `SDW1-Playback` and `SDW1-Capture` dailinks, we will have
`SDW0-Playback` dailink .init = sof_sdw_rtd_init which will call rt_sdca_jack_rtd_init and set `rt712-sdca-aif1` dai.rtd_init_done = true
`SDW0-Capture` dailink .init = sof_sdw_rtd_init which will call nothing since `rt712-sdca-aif1` dai.rtd_init_done is true.
`SDW1-Playback` dailink .init = sof_sdw_rtd_init which will call `rt_amp_spk_rtd_init` once since and 
`rt712_spk_rtd_init`
`SDW1-Capture` dailink .init = sof_sdw_rtd_init which will call nothing since `rt712-sdca-aif1` dai.rtd_init_done and `rt1316-aif` dai.rtd_init_done are both true.

charleskeepax replied:
Are we confident that we only need to call rtd_init once? I could well imagine a situation, where the same device could be used in two different DAI links, thus requiring different slightly different init.

> it goes back to the change we made: all instances of a dai need to be in the same rtd/dailink.
> 
> We don't support e.g. two headset codecs used independently, or two identical amplifiers handling different content on separate links.

Do you have any more info on what change this was? I am curious as to what is introducing these limitations.

bardliao replied:
> Are we confident that we only need to call rtd_init once? I could well imagine a situation, where the same device could be used in two different DAI links, thus requiring different slightly different init.

Yes, we use predefined be_id, all amp dai link will use the same BE id. IOW, we can't support amps on different dai links on the existing code. That is a restriction of the machine driver.

> 
> > it goes back to the change we made: all instances of a dai need to be in the same rtd/dailink.
> > We don't support e.g. two headset codecs used independently, or two identical amplifiers handling different content on separate links.
> 
> Do you have any more info on what change this was? I am curious as to what is introducing these limitations.

The change was to add the capability to support different codec models. For example, say if Cohen supports amp, and customers want to use Cohen's amp + Jamerson work together in the same dailnik. (4 speakers, 2 on Cohen and 2 on Jamerson).

charleskeepax replied:
Ah I see.

plbossart reviewed (commented):
Two comments on the use of the 'initialized' field, this is a bit controversial and potentially problematic @bardliao

plbossart reviewed (dismissed):
thanks @bardliao

RanderWang commented on sound/soc/intel/boards/sof_board_helpers.c:
```diff
@@ -248,6 +248,24 @@ int sof_intel_board_set_intel_hdmi_link(struct device *dev,
+		for_each_rtd_codec_dais(rtd, i, dai)
+			if (strstr(dai->name, dai_name[index])) {
+				dev_dbg(rtd->card->dev, "get dai %s\n", dai->name);
+				return dai;
```
so return when  the first code dai is matched, how about others ?

bardliao replied:
> so return when the first code dai is matched, how about others ?

The purpose is to get the specific codec dai in the dailink and return it. No need to care about other dais even they may belong to the same codec.

RanderWang commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -1371,6 +1393,57 @@ static void set_dailink_map(struct snd_soc_dai_link_codec_ch_map *sdw_codec_ch_m
+			if (ret)
+				return ret;
+		}
+		codec_info->dais[dai_index].rtd_init_done = true;
```
When do you clear rtd_init_done ?  It is saved in static codec_info_list, so if we remove  audio card and probe card again, we will find it is done ?

bardliao replied:
> When do you clear rtd_init_done ? It is saved in static codec_info_list, so if we remove audio card and probe card again, we will find it is done ?

Thanks @RanderWang for pointing this out. I clear them in `mc_dailink_exit_loop` now.

plbossart commented:
merge conflicts @bardliao

bardliao commented:
> merge conflicts @bardliao

Fixed

plbossart commented:
quite a few checkpatch warnings @bardliao, is it worth fixing?

bardliao commented:
> quite a few checkpatch warnings @bardliao, is it worth fixing?

Yes, and fixed. Thanks @plbossart

plbossart commented:
CI shows this error 

rmmod: ERROR: Module snd_soc_intel_hda_dsp_common is in use by: snd_soc_intel_sof_board_helpers

bardliao commented:
> CI shows this error
> 
> rmmod: ERROR: Module snd_soc_intel_hda_dsp_common is in use by: snd_soc_intel_sof_board_helpers

We need to remove snd_soc_intel_sof_board_helpers first. https://github.com/thesofproject/sof-test/pull/1131

bardliao commented:
> > CI shows this error
> > rmmod: ERROR: Module snd_soc_intel_hda_dsp_common is in use by: snd_soc_intel_sof_board_helpers
> 
> We need to remove snd_soc_intel_sof_board_helpers first. [thesofproject/sof-test#1131](https://github.com/thesofproject/sof-test/pull/1131)

Ah, https://github.com/thesofproject/sof-test/pull/1130/ includes it.

plbossart commented:
sorry @bardliao now we have a conflict with @brentlu's changes. D'oh.

bardliao commented:
> sorry @bardliao now we have a conflict with @brentlu's changes. D'oh.

Fixed

plbossart commented:
SOFCI TEST

plbossart commented:
SOFCI TEST

bardliao commented:
SOFCI TEST

marc-hb commented:
Just for the record, regression:
- #4834 

Fix submitted:
- #4835
